File: src/main/java/no/novari/kafka/topic/name/TopicNameMappingService.java
package no.novari.kafka.topic.name;

import org.springframework.stereotype.Service;

import java.util.Objects;
import java.util.StringJoiner;

@Service
final class TopicNameMappingService {

    String toTopicName(TopicNameParameters topicNameParameters) {

        StringJoiner topicNameJoiner = new StringJoiner(".")
                .add(topicNameParameters.getTopicNamePrefixParameters().getOrgId())
                .add(topicNameParameters.getTopicNamePrefixParameters().getDomainContext())
                .add(topicNameParameters.getMessageType().getTopicNameParameter());

        topicNameParameters.getTopicNameParameters()
                .stream()
                .map(TopicNameParameter::getValue)
                .filter(Objects::nonNull)
                .forEachOrdered(topicNameJoiner::add);

        return topicNameJoiner.toString();
    }

}

File: src/main/kotlin/no/novari/kafka/topic/name/TopicNameMappingService.kt
package no.novari.kafka.topic.name

import org.springframework.stereotype.Service
import java.util.StringJoiner

@Service
internal class TopicNameMappingService {
    fun toTopicName(topicNameParameters: TopicNameParameters): String {
        val topicNameJoiner =
            StringJoiner(".")
                .add(topicNameParameters.topicNamePrefixParameters.orgId)
                .add(topicNameParameters.topicNamePrefixParameters.domainContext)
                .add(topicNameParameters.messageType.topicNameParameter)

        topicNameParameters.topicNameSuffixParameters
            .mapNotNull { it.value }
            .forEach { topicNameJoiner.add(it) }

        return topicNameJoiner.toString()
    }
}
